docs/architecture/widget_framework.md

```diff
@@ -0,0 +1,90 @@
+
+Each screen's content is composed entirely of widgets, and when a screen requests new data, no matter which type it is, the same backend logic is used to look up the screen's configuration, fetch the appropriate real-time data, transform that data into "candidate" widgets, and then assemble some of those candidates together to be sent back to the client and rendered.
+
+## What's a widget?
+
+A widget is a self-contained piece of content. It's essentially a special type of Model (as in MVC) that can be dropped into template slots alongside any number of other widgets to form a full page of content.
+
+Some examples of widgets include the departures list, subway status, alert cards, and image PSAs. However, even basic screen content like the header and footer are widgets as well.
+
+On the backend, each widget implements the [`WidgetInstance` protocol](/lib/screens/v2/widget_instance.ex). The framework calls these protocol functions to find out each widget's priority and where it would like to appear on the screen. The framework then determines how best to position all available widgets within these constraints.
+
+**Widgets do not directly communicate with one another, or even know about the existence of other widgets.** Widgets only communicate with the framework via the aforementioned protocol functions. Once created, a widget has all of the data it needs to render itself.
+
+## Why widgets?
+
+We have many screens, and many different _types_ of screens. Each one needs to display the most rider-relevant info for its specific location.
+
+The widget framework allows us to automate the process of deciding what content is most relevant from moment to moment on a given screen.
+
+The strict isolation of widgets prevents an explosion in complexity and interdependence. It also makes it possible to swap pieces of content out on a screen without worrying about potential side-effects. Business logic (of which there is much since the T has endless edge cases) is corralled inside `WidgetInstance`-implementing modules.
+
+Once a widget is created, it has all of the data it needs. This means all functions on a widget are pure/free of side effects, and easy to test.
+
+See the [`WidgetInstance` protocol](/lib/screens/v2/widget_instance.ex)'s inline documentation for details on each implementation function.
+
+## Request data flow / how the framework logic works
+
+Screens team members can see this as a [flowchart in Miro](https://miro.com/app/board/o9J_lDRaax4=/).
+
+1. Client requests `/v2/api/screen/{id}`. Subsequent steps take place on the server.
+1. We get the config stored under `id` in the screen configuration file.
+1. We look up the "app ID" stored in the config. We get the appropriate candidate generator module for that app ID.
+1. We get the template for this screen type by calling `candidate_generator.screen_template()`.
+1. We get a list of candidate widgets for the screen by calling `candidate_generator.candidate_instances(config)`.
```

Comment:
Somewhere separate, it could be helpful to acknowledge that there are often more candidates than space on screens, and we weigh which ones should appear based on priority. Examples would also be helpful: alert widget candidates are highly prioritized and so almost always make it onto screens, while general PSAs are low-priority and will get bumped if the screen template runs out of room.